src/token_bucket.rs

```diff
@@ -3,6 +3,45 @@ use std::time::{Duration, Instant};
 
 use crate::error::Error;
 
+/// Implementation of the [token bucket](https://en.wikipedia.org/wiki/Token_bucket)
+/// rate-limiting algorithm.
+///
+/// The algorithm is based on the analogy with a fixed capacity *bucket* to
+/// which abstract *tokens* are added with a constant rate up until the bucket
+/// is full.
+///
+/// To implement rate-limiting, each request entering the system is assigned
+/// a *cost* in terms of the number of tokens, and it tries to consume that
+/// many tokens from the bucket:
+///
+/// * if the bucket has the sufficient number of tokens available, they are *consumed*,
+///   and the request is allowed to go through
+///
+/// * otherwise, the request is rejected. New requests continue to be rejected until
+///   the bucket is refilled
+///
+/// Because tokens are added to the bucket with a constant rate over the specified
+/// interval of time, this effectively limits requests to the system to the same rate.
+///
+/// Bucket capacity and token replenishment rate are specified at bucket creation time.
+///
+/// ```
+/// use std::time::Duration;
+/// use youshallnotpass::{TokenBucket, Error};
+///
+/// // create a bucket that allows to consume 3 tokens every 60 seconds
+/// let bucket = TokenBucket::new(3, Duration::from_secs(60));
+/// assert!(bucket.consume(1).is_ok());
+/// assert!(bucket.consume(1).is_ok());
+/// assert!(bucket.consume(1).is_ok());
+/// // requests exceeding the configured rate-limit are rejected. The error will specify
+/// // how much time the caller has to wait before retrying
+/// assert!(matches!(bucket.consume(1), Err(Error::RetryAfter(duration))));
+/// ```
+///
+/// Generated tokens can be consumed all at once or over time.
+///
+/// This type implements [`Sync`] and can be shared between threads.
```

Comment:
Q: Does it make sense to specify it explicitly? Can't it be inferred automatically in docs?

Reply:
It is redundant yes. Let me drop it.